File: Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        // [][0] for buy, [][1] for sell
        if (prices.length <= 1)
            return 0;
        int ans = 0;
        int[][] dp = new int[prices.length][2];

        for (int i = 1; i < prices.length; i++) {
            for (int j = i - 1; j >= 0; j--) {
                dp[i][1] = Integer.max(dp[i][1], dp[j][0] + prices[i] - prices[j]);
                if (j <= i - 2)
                    dp[i][0] = Integer.max(dp[i][0], dp[j][1]);
            }
            ans = Integer.max(ans, Integer.max(dp[i][1], dp[i][0]));
        }

        return ans;
    }
}

class Best_Time_to_Buy_and_Sell_Stock_with_Cooldown {
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[][] testset = { { 1, 2, 3, 0, 2 }, { 1, 2, 1, 100, 1, 2, 3 } };
        for (int[] prices : testset)
            System.out.println(sol.maxProfit(prices));
    }
}

File: Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.cpp
#include <iostream>
#include <vector>
using namespace std;
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/discuss/75928/Share-my-DP-solution-(By-State-Machine-Thinking)
class Solution {
  public:
    int maxProfit(vector<int> &prices) {
        if (prices.size() <= 1)
            return 0;
        vector<int> s0(prices.size());
        vector<int> s1(prices.size());
        vector<int> s2(prices.size());
        s1[0] = -prices[0];
        for (int i = 1; i < prices.size(); i++) {
            s0[i] = max(s0[i - 1], s2[i - 1]);
            s1[i] = max(s1[i - 1], s0[i - 1] - prices[i]);
            s2[i] = s1[i - 1] + prices[i];
        }
        return max(s0[prices.size() - 1], s2[prices.size() - 1]);
    }
};

int main(int argc, char const *argv[]) {
    Solution sol;
    vector<vector<int>> testset = {{1, 2, 3, 0, 2}, {1, 2, 1, 100, 1, 2, 3}};
    for (auto prices : testset)
        cout << sol.maxProfit(prices) << endl;

    return 0;
}
